src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/SafeHandles.Linux.cs

```diff
@@ -0,0 +1,74 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Win32.SafeHandles;
+
+namespace System.DirectoryServices.Protocols
+{
+    internal sealed class ConnectionHandle : SafeHandleZeroOrMinusOneIsInvalid
+    {
+        internal bool _needDispose = false;
+
+        internal ConnectionHandle()
+            :base(true)
+        {
+            OpenLDAP.ldap_initialize(out IntPtr ldap, null);
+            SetHandle(ldap);
+        }
+
+        internal ConnectionHandle(IntPtr value, bool disposeHandle) : base(true)
+        {
+            _needDispose = disposeHandle;
+            if (value == IntPtr.Zero)
+            {
+                throw new LdapException("There was an error when attempting to initialize a connection with the LDAP server");
+            }
+            else
+            {
+                SetHandle(value);
+            }
+        }
+        protected override bool ReleaseHandle()
+        {
+            if (handle != IntPtr.Zero)
```

Comment:
Because you're SafeHandleZeroOrMinusOneIsInvalid, IsInvalid will already report true for handle being IntPtr.Zero; so this if can't evaluate both ways.

This whole body can just be

```C#
IntPtr nullPointer = IntPtr.Zero;
int ret = OpenLDAP.ldap_unbind_ext_s(handle, ref nullPointer, ref nullPointer);
handle = IntPtr.Zero;

// Assuming you do want to fail for any error from unbind.
// Note that returning false from here will terminate the process.
return ret == 0;
```

Reply:
Is it at all possible for ReleaseHandle to get called twice? I thought that is what this was protecting against similar to when you also protect in case double Dispose gets called.

Reply:
Nope, SafeHandle already guarantees that.  (Well, I suppose you could make a public method which directly calls ReleaseHandle, but nothing within SafeHandle will do it)